myia/operations/macro_to_scalar_type.py

```diff
@@ -0,0 +1,55 @@
+"""
+Implementation of 'to_scalar_type' operation.
+
+Convert given data into an appropriate abstract scalar.
+If data is already an abstract type, return the result of method xvalue().
+If data is already an abstract scalar with a number type, return it as is.
+If data is an abstract scalar with a string type, try to infer type
+from given string and return an abstract scalar with inferred type.
+Used in myia/operations/op_full.py
+"""
+
+import numpy as np
+
+from ..lib import (
+    ANYTHING,
+    TYPE,
+    VALUE,
+    AbstractScalar,
+    AbstractType,
+    Constant,
+    macro,
+)
+from ..xtype import Number, String, pytype_to_myiatype
+
+
+@macro
+async def to_scalar_type(info, data):
+    """
+    Convert given data to abstract scalar.
+
+    :param data: arbitrary data to convert
+    :return: an abstract scalar object if data can be converted,
+        otherwise raise an exception.
+    """
+    sync_data = await data.get()
+    if isinstance(sync_data, AbstractType):
+        sync_data = sync_data.xvalue()
+    if isinstance(sync_data, AbstractScalar):
+        xtype = sync_data.xtype()
+        if issubclass(xtype, Number):
```

Comment:
There are two issues here:

1. This will raise a TypeError if xtype is a future. One of your tests triggers this (I put a comment under it). The test still passes, but that's because it is expecting the other TypeError, the one raised by this function.
2. If you do force the future/pending value, then `to_scalar_type(10)` will work, which it shouldn't.